### PR TITLE
feat(shell): implement support for background jobs

### LINKED ISSUE #1349: [Bug] Scripts don't accept & operator
**Describe the bug**

In order to run cypress tests, I use the & operator to background the execution of the process as described [here](https://docs.cypress.io/guides/guides/continuous-integration.html#Boot-your-server). This worked fine in Yarn 1, but in Yarn 2 if any script contains this operator I get

```
Syntax Error: Expected "$", "$(", "${", "&&", "'", ";", "<", "<<<", ">", ">>", "\"", "\\", "|", "|&", "||", [ \t], or end of input but "&" found (line 1, column 10).
    at tt (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:28:320521)
    at Object.parse (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:28:330627)
    at Object.t.parseShell (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:28:315544)
    at Object.t.execute (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:16:19878)
    at A.configuration.reduceHook.script (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:16:5595)
    at /Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:16:5724
    at async s.mktempPromise (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:2:5648)
    at async Object.B [as executePackageScript] (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:16:5355)
    at async c.execute (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:50:58654)
    at async c.validateAndExecute (/Users/wainwb01/repos/planv7/.yarn/releases/yarn-berry.js:16:50176)   
```
                                                                          
**To Reproduce**
Add "&" to any "script" in your package.json then run the script

### PR DESCRIPTION
**Pull request by paul-soporan**
**What's the problem this PR addresses?**

This PR implements support for background jobs in Yarn's portable shell.

Closes #1349.

**How did you fix it?**

We execute the jobs in the original process as the shell state isn't fully serializable to be sent between processes. This means that there's no way to "detach" the job and do something else in the terminal tab at the same time, but it also has the advantage that you can `CTRL + C` the entire process and the background job will be killed (if we used detached processes, once all foreground jobs finished there would be no way to kill the background jobs without the `kill` command).

(TTY only) We prefix the output of background jobs with color-coded `[${index}]` prefixes to make it easier to tell which lines come from which job (since everything is interlaced).

(TTY only) We also show a "job has ended" message after each background job ends, similar to Fish. It includes that same color-coded prefix and it also includes the stringified command chain that was executed.

This PR implements the ZSH and Fish behavior (having `&` only affect the last chain in a line) rather than the Bash behavior (having `&` affect all chains in the line). In other words, `sleep 3 && echo foo & echo bar` is the same as `sleep 3 && {echo foo & echo bar}`, the other possibility being `{sleep 3 && echo foo} & echo bar`.

**Checklist**

- [X] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [X] I have set the packages that need to be released for my changes to be effective.

- [X] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-shell/sources/index.ts [hunk 1]
```diff
@@ -688,11 +690,41 @@ async function executeCommandChain(node: CommandChain, opts: ShellOptions, state
...
+
+  if (background) {
+    const index = state.nextBackgroundJobIndex++;
+    const colorizer = getColorizer(index);
+    const rawPrefix = `[${index}]`;
+    const prefix = colorizer(rawPrefix);
+
+    const {stdout, stderr} = createOutputStreamsWithPrefix(state, {prefix});
+
+    executeCommandChainImpl(node, opts, cloneState(state, {stdout, stderr}))
```
Comment (arcanis) [inline packages/yarnpkg-shell/sources/index.ts:709]: How come we never `await` this promise anywhere? Could there be a race condition risk in the tests, where the "background" jobs aren't finished by the time `bufferResult` returns?

Comment (paul-soporan) [reply packages/yarnpkg-shell/sources/index.ts:709]: I intended to store these promises and await them right before `executeShellLine` finishes (i.e. `echo $(sleep 10 & echo foo)` should still sleep for 10 seconds) but I didn't include it in the original prototype and forgot about it before opening this PR, thanks for noticing!

File: packages/yarnpkg-shell/tests/shell.test.ts [hunk 2]
```diff
@@ -1566,4 +1562,35 @@ describe(`Shell`, () => {
       });
     });
   });
+
+  describe(`Background jobs`, () => {
+    it(`should provide color-coded prefixed output and color-coded "Job has ended" message inside TTYs`, async () => {
```
Comment (arcanis) [inline packages/yarnpkg-shell/tests/shell.test.ts:1567]: Can you add a test doing an actual `sleep` (perhaps implemented as a builtin in tests only)? This way we'd also guard against race conditions like the one there used to be.

**Review by arcanis — APPROVED** [review]
One extra test then good to go 👍

**Pull request merged**